talk/morelanguage/exceptions.tex

```diff
@@ -45,35 +45,164 @@
+        \item prefer standard exception classes
       \end{itemize}
-      \item an exception will be caught if the type in the catch clause matches or is a base class of the thrown object's static type
+      \item throw objects by value
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+    void process_data(file& f) {
+      if (!f.open())
+        throw std::invalid_argument{"stream is not open"};
+
+      auto header = read_line(f); // may throw an IO error
+      if (!header.starts_with("BEGIN"))
+        throw std::runtime_error{"invalid file content"};
+
+      std::string body(f.size()); // may throw std::bad_alloc
+      ...
+    }
+  \end{cppcode}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Exceptions}
+  \begin{block}{Catching exceptions}
+    \begin{itemize}
+      \item a catch clause catches an exception of the same or derived type
+      \item multiple catch clauses will be matched in order
+      \item if no catch clause matches, the exception propagates
+      \item if the exception is never caught, \mintinline{cpp}{std::terminate} is called
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+    try {
+      process_data(f);
+    } catch (const std::invalid_argument& e) {
+      bad_files.push_back(f);
+    } catch (const std::exception& e) {
+      std::cerr << "Failed to process file: " << e.what();
+    }
+  \end{cppcode}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Exceptions}
+  \begin{block}{Rethrowing exceptions}
+    \begin{itemize}
+      \item a caught exception can be rethrown
+      \item useful when we want to act on an error, but cannot handle and want to propagate it
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+    try {
+      process_data(f);
+    } catch (const std::bad_alloc& e) {
+      std::cerr << "Insufficient memory for " << f.name();
+      throw; // rethrow
+    }
+  \end{cppcode}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Exceptions}
+  \begin{block}{Catching everything}
+    \begin{itemize}
+    \item sometimes we need to catch all possible exceptions
+    \item e.g.\ in \mintinline{cpp}{main}, a thread, a destructor, interfacing with C, \ldots
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+
+    try {
+      callUnknownFramework();
+    } catch(const std::exception& e) {
+      // catches std::exception and all derived types
+      std::cerr << "Exception: " << e.what() << std::endl;
+    } catch(...) {
+      // catches everything else
+      std::cerr << "Unknown exception type" << std::endl;
+    }
+  \end{cppcode}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Exceptions}
+  \begin{block}{Stack unwinding}
+    \begin{itemize}
+      \item all objects on the stack between a \mintinline{cpp}{throw} and the matching \mintinline{cpp}{catch} are destructed automatically
+      \item this should cleanly release intermediate resources
+      \item make sure you are using the RAII idiom for your own classes
```

Comment:
This will only come later in the lecture. I don't think it makes sense to link there, though.